### PR TITLE
 Android Gradle Plugin upgrade to v7.3.0 

### LINKED ISSUE #8: Upgrade gradle version
copy of https://github.com/microsoft/rural-crowdsourcing-toolkit/issues/136

### PR DESCRIPTION
**Pull request by thedroiddiv**
## Purpose / Description
Upgrades gradle to the latest version and fix compatibility issues with other libraries that have been upgraded along with the gradle. More descriptions has been left in code reviews.

## Fixes
Fixes #8 
Fixes https://github.com/microsoft/rural-crowdsourcing-toolkit/issues/136

## How Has This Been Tested?
All libraries are compatible and project builds as expected.

## Checklist

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: app/build.gradle.kts [hunk 1]
```diff
@@ -3,20 +3,20 @@ import com.android.build.gradle.internal.cxx.configure.gradleLocalProperties
...
     id("com.google.firebase.crashlytics")
-    id("dagger.hilt.android.plugin")
+    id(Plugin.Hilt.id) // version is picked up from parent i.e project'd root build.gradle
     id("androidx.navigation.safeargs.kotlin")
     id("com.github.ben-manes.versions") version "0.38.0"
 }
 
 android {
-    compileSdkVersion(31)
+    compileSdk = 33
```
Comment (thedroiddiv) [inline app/build.gradle.kts:15]: setter functions are deprecated and new way is using property access syntax

File: app/build.gradle.kts [hunk 2]
```diff
@@ -41,19 +41,19 @@ android {
         }
     }
     compileOptions {
-        sourceCompatibility = JavaVersion.VERSION_1_8
-        targetCompatibility = JavaVersion.VERSION_1_8
+        sourceCompatibility = JavaVersion.VERSION_11
+        targetCompatibility = JavaVersion.VERSION_11
```
Comment (thedroiddiv) [inline app/build.gradle.kts:45]: a requirement for gradle 7.0.0 onwards

File: app/build.gradle.kts [hunk 3]
```diff
@@ -133,11 +134,8 @@ dependencies {
     implementation(Dependencies.Google.Firebase.crashlytics)
     implementation(Dependencies.Google.Firebase.analytics)
 
-    implementation(Dependencies.AndroidX.Hilt.dagger)
-    implementation(Dependencies.AndroidX.Hilt.hiltNavigationFragment)
-
-    kapt(Dependencies.AndroidX.Hilt.daggerCompiler)
-    kapt(Dependencies.AndroidX.Hilt.daggerHiltCompiler)
+    implementation(Dependencies.Google.Hilt.hiltAndroid)
+    kapt(Dependencies.Google.Hilt.hiltAndroidCompiler)
```
Comment (thedroiddiv) [inline app/build.gradle.kts:138]: Hilt can now be independently used without additional dagger dependencies.

File: app/build.gradle.kts [hunk 4]
```diff
@@ -183,3 +181,7 @@ dependencies {
     "largeImplementation" ("com.google.mlkit:face-detection:16.1.2")
     "largeImplementation" ("com.github.fishwjy:VideoCompressor:master-SNAPSHOT")
 }
+
+kapt {
+    correctErrorTypes = true
+}
```
Comment (thedroiddiv) [inline app/build.gradle.kts]: a configuration for hilt

File: app/src/main/AndroidManifest.xml [hunk 5]
```diff
@@ -23,6 +23,7 @@
 
         <activity
             android:name=".ui.MainActivity"
+            android:exported="true"
```
Comment (thedroiddiv) [inline app/src/main/AndroidManifest.xml:26]: requirement for android SDK 33

File: app/src/main/java/com/microsoft/research/karya/ui/assistant/Assistant.kt [hunk 6]
```diff
@@ -80,19 +80,19 @@ constructor(
     }
 
     @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
-    private fun stopAssistant() {
+    fun stopAssistant() {
```
Comment (thedroiddiv) [inline app/src/main/java/com/microsoft/research/karya/ui/assistant/Assistant.kt:83]: newer version of the library (lifecycle) requires to be public

File: build.gradle.kts [hunk 7]
```diff
@@ -16,9 +16,9 @@ buildscript {
         }
     }
 
+    // TODO: Move to plugins block (recommended from AGP 7.3)
     dependencies {
         classpath(Plugins.agp)
```
Comment (thedroiddiv) [inline build.gradle.kts:21]: newer version of gradle relies mostly on plugin mechanism, good to move other as well later but for now only hilt is required

File: buildSrc/src/main/java/Dependencies.kt [hunk 8]
```diff
@@ -1,25 +1,29 @@
...
   const val kotlin = "org.jetbrains.kotlin:kotlin-gradle-plugin:$KOTLIN_VERSION"
-  const val hilt = "com.google.dagger:hilt-android-gradle-plugin:$DAGGER_HILT_VERSION"
   const val safeArgs = "androidx.navigation:navigation-safe-args-gradle-plugin:$NAVIGATION_VERSION"
 }
 
+// TODO: Rename to Plugins once root build.gradle is migrated from using classpath to plugins
+sealed class Plugin(val id: String, val version: String) {
+  object Hilt : Plugin("com.google.dagger.hilt.android", HILT_VERSION)
+  object KtLint : Plugin("org.jlleitschuh.gradle.ktlint", KT_LINT_VERSION)
+}
```
Comment (thedroiddiv) [inline buildSrc/src/main/java/Dependencies.kt:25]: from classpaths to plugin, a better way

File: buildSrc/src/main/java/Dependencies.kt [hunk 9]
```diff
@@ -45,23 +49,14 @@ object Dependencies {
...
     const val datastorePrefs = "androidx.datastore:datastore-preferences:1.0.0-beta01"
 
-    object Hilt {
-      const val dagger = "com.google.dagger:hilt-android:$DAGGER_HILT_VERSION"
-      const val daggerCompiler = "com.google.dagger:hilt-compiler:$DAGGER_HILT_VERSION"
-      const val daggerHiltCompiler = "androidx.hilt:hilt-compiler:$ANDROIDX_HILT_VERSION"
-      const val hiltLifecycleViewModel =
-          "androidx.hilt:hilt-lifecycle-viewmodel:$ANDROIDX_HILT_VERSION"
-      const val hiltNavigation = "androidx.hilt:hilt-navigation:$ANDROIDX_HILT_VERSION"
-      const val hiltNavigationFragment = "androidx.hilt:hilt-navigation-fragment:1.0.0-beta01"
```
Comment (thedroiddiv) [inline buildSrc/src/main/java/Dependencies.kt:55]: Hilt is not a part of AndroidX but rather a Google library, so moved to Google block

**Review by anuragshukla06 — CHANGES_REQUESTED** [review]
While you are at it, can you please shift the dependencies that have been directly mentioned in the gradle file to the Dependecies file? ;-;

**Comment by thedroiddiv** [comment]
When done just let me know, and I'll squash the last two commits. Left it now so can be easily reviewed.

**Review by anuragshukla06 — APPROVED** [review]

**Pull request merged**